tests/bluetooth/tester/src/btp_has.c

```diff
@@ -0,0 +1,204 @@
+/* btp_has.c - Bluetooth HAS Tester */
+
+/*
+ * Copyright (c) 2023 Oticon
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/bluetooth/audio/has.h>
+
+#include "btp/btp.h"
+#include "zephyr/sys/byteorder.h"
+#include <stdint.h>
+
+#include <zephyr/logging/log.h>
+#define LOG_MODULE_NAME bttester_has
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_BTTESTER_LOG_LEVEL);
+
+static uint8_t has_supported_commands(const void *cmd, uint16_t cmd_len,
+				       void *rsp, uint16_t *rsp_len)
+{
+	struct btp_has_read_supported_commands_rp *rp = rsp;
+
+	/* octet 0 */
+	tester_set_bit(rp->data, BTP_HAS_READ_SUPPORTED_COMMANDS);
+
+	*rsp_len = sizeof(*rp) + 1;
+
+	return BTP_STATUS_SUCCESS;
+}
+
+static uint8_t has_set_active_index(const void *cmd, uint16_t cmd_len,
+			   void *rsp, uint16_t *rsp_len)
+{
+	const struct btp_has_set_active_index_cmd *cp = cmd;
+	/*
+	struct bt_conn *conn;
+
+	conn = bt_conn_lookup_addr_le(BT_ID_DEFAULT, &cp->address);
+	if (!conn) {
+		LOG_ERR("Unknown connection");
+		return BTP_STATUS_FAILED;
+	}
+	*/
+	int err = bt_has_preset_active_set(cp->index);
+	return (err) ? BTP_STATUS_FAILED : BTP_STATUS_SUCCESS;
+}
+
+static uint8_t has_presets = 0;
+static char    temp_name[BT_HAS_PRESET_NAME_MAX+1];
+
+static uint8_t has_set_preset_name(const void *cmd, uint16_t cmd_len,
+			   void *rsp, uint16_t *rsp_len)
+{
+	const struct btp_has_set_preset_name_cmd *cp = cmd;
+
+	memcpy(temp_name, cp->name, cp->length);
+	temp_name[cp->length] = '\0';
+	int err = bt_has_preset_name_change(cp->index, temp_name);
+	return (err) ? BTP_STATUS_FAILED : BTP_STATUS_SUCCESS;
+}
+
+static uint8_t bit_position(uint8_t value)
+{
+	uint8_t position = 0;
+
+	while (value) {
+		value >>= 1;
+		position++;
+	}
+	return position;
+}
+
+static uint8_t has_remove_preset(const void *cmd, uint16_t cmd_len,
+			   void *rsp, uint16_t *rsp_len)
+{
+	const struct btp_has_remove_preset_cmd *cp = cmd;
+	int err = 0;
+
+	if (BT_HAS_PRESET_INDEX_NONE == cp->index) {
+		uint8_t index;
+
+		while (has_presets) {
+			index = bit_position(has_presets & -has_presets);
+			if (index) {
+				err = bt_has_preset_unregister(index);
+				has_presets &= ~(1<<(index-1));
+			}
+			else {
+				break;
+			}
+		}
+	}
+	else {
+		err = bt_has_preset_unregister(cp->index);
+	}
+	return (err) ? BTP_STATUS_FAILED : BTP_STATUS_SUCCESS;
+}
+
+static int has_preset_selected(unsigned char index, bool sync)
+{
+	return BTP_STATUS_SUCCESS;
+}
+
+static const struct bt_has_preset_ops has_preset_ops = { has_preset_selected, NULL };
+
+static uint8_t has_add_preset(const void *cmd, uint16_t cmd_len,
+			   void *rsp, uint16_t *rsp_len)
+{
+	const struct btp_has_add_preset_cmd *cp = cmd;
+
+	memcpy(temp_name, cp->name, cp->length);
+	temp_name[cp->length] = '\0';
```

Comment:
Missing check for if `cp->length > sizeof(temp_name)`

Reply:
Done.

Reply:
Looks like it's fixed in `has_set_preset_name`, but not here. Maybe we should define a setter-function?